models/intermediate/int_transform_contract_code.sql

```diff
@@ -0,0 +1,175 @@
+            , n_table_entries
+            , n_types
+            , n_data_segments
+            , n_elem_segments
+            , n_imports
+            , n_exports
+            , n_data_segment_bytes
+            , batch_id
+            , batch_run_date
+            , cast('{{ execution_date }}' as timestamp) as airflow_start_ts
+            -- Rank records to identify the latest entry
+            , row_number() over (
+                partition by ledger_key_hash, cast(closed_at as date)
+                order by closed_at desc, ledger_sequence desc
+            ) as row_num
+        from {{ ref('stg_contract_code') }}
+        where
+            -- Process only completed days up to execution_date
+            date(closed_at) < date('{{ execution_date }}')
+    )
+
+    , filtered_data as (
+        select *
+        from source_data
+        where row_num = 1  -- Only consider the latest record
+    )
+
+    , derived_data as (
+        select
+            ledger_key_hash
+            , min(case when ledger_entry_change = 0 then closed_at end) as contract_create_ts  -- Calculate contract created timestamp
+            -- Calculate contract deletion timestamp
+            , max(case when ledger_entry_change = 2 and deleted = true then closed_at end) as contract_delete_ts
+        from source_data
+        group by ledger_key_hash
+    )
+
+    , transformed_data as (
+        select
+            f.ledger_key_hash
+            , f.contract_code_hash
+            , f.last_modified_ledger
+            , f.ledger_entry_change
+            , f.ledger_sequence
+            , f.deleted
+            , f.closed_at
+            , f.n_instructions
+            , f.n_functions
+            , f.n_globals
+            , f.n_table_entries
+            , f.n_types
+            , f.n_data_segments
+            , f.n_elem_segments
+            , f.n_imports
+            , f.n_exports
+            , f.n_data_segment_bytes
+            , d.contract_create_ts
+            , d.contract_delete_ts
+            , f.airflow_start_ts
+            , f.batch_id
+            , f.batch_run_date
+            , sha256(concat(
+                coalesce(f.ledger_key_hash, '')
+                , coalesce(cast(f.n_instructions as string), '')
+                , coalesce(cast(f.n_functions as string), '')
+                , coalesce(cast(f.n_globals as string), '')
+                , coalesce(cast(f.n_table_entries as string), '')
+                , coalesce(cast(f.n_types as string), '')
+                , coalesce(cast(f.n_data_segments as string), '')
+                , coalesce(cast(f.n_elem_segments as string), '')
+                , coalesce(cast(f.n_imports as string), '')
+                , coalesce(cast(f.n_exports as string), '')
+                , coalesce(cast(f.n_data_segment_bytes as string), '')
+                , coalesce(cast(d.contract_create_ts as string), '')
+                , coalesce(cast(d.contract_delete_ts as string), '')
+            )) as row_hash
```

Comment:
Is `row_hash` needed for unique_id? Isn't `ledger_key_hash` effectively the unique_id for an entry and then `ledger_key_hash` + `closed_at` the unique row?